### PR TITLE
Feat/nnx ckpt merge state refactor

### PR DESCRIPTION
**Pull request by ecnal-cienet**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

You can also provide a comma-separated list. If you don't want to close a bug but
simply to reference it, use BUGS, e.g.: 
BUGS: b/123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/maxtext_utils.py [hunk 1]
```diff
@@ -1501,44 +1501,62 @@ def setup_initial_state(
...
+        # The restore already overlaid the checkpoint onto a copy of the abstract, so a leaf it
+        # didn't carry is still an unmaterialized placeholder. Fill those from the fresh init: a
+        # present leaf comes from the checkpoint, an absent one keeps its init value.
+        overlay = restored if is_emergency else restored["items"]
+        merged = jax.tree.map(
+            lambda ckpt, init: init if isinstance(ckpt, jax.ShapeDtypeStruct) else ckpt,
+            overlay.to_pure_dict(),
+            state.to_pure_dict(),
+            is_leaf=lambda x: isinstance(x, jax.ShapeDtypeStruct),
+        )
```
Comment (xibinliu) [inline src/maxtext/utils/maxtext_utils.py:1534]: get the "rest" part from state (values), and replace the rest part in overlay (abstract rest)

common function for getting rest?

File: src/maxtext/common/checkpointing.py
Comment (xibinliu) [inline src/maxtext/common/checkpointing.py:953]: add a TODO: check the _enforce_missing_weight_policy for the raw params

File: src/maxtext/common/train_state_nnx.py [hunk 2]
```diff
@@ -193,3 +199,25 @@ def from_linen_checkpoint_dict(linen_pure_dict):
...
+  """Reshapes an nnx.State into the on-disk checkpoint layout, splitting by Variable type.
+
+  Weights (nnx.Param) and the optimizer state map to the Linen
+  params/opt_state/step layout, so pure_nnx and Linen checkpoints stay
+  interchangeable. rngs/dropout (nnx.RngState) and batch stats (nnx.BatchStat) go
+  under an `nnx_aux` subtree so they survive a resume. Works on a concrete state
+  (save) or an abstract state (restore target).
+  """
+  params, rng_state, batch_stats, rest = nnx.split_state(state, nnx.Param, nnx.RngState, nnx.BatchStat, ...)
+  # `rest` holds the optimizer state (mu/nu/step aren't nnx.Param), which must be
```
Comment (xibinliu) [inline src/maxtext/common/train_state_nnx.py:214]: nit: nnx.split_state(state.model, ....) ?

File: src/maxtext/common/train_state_nnx.py [hunk 3]
```diff
@@ -193,3 +199,25 @@ def from_linen_checkpoint_dict(linen_pure_dict):
   if optimizer:
     result["optimizer"] = optimizer
   return result
+
+
+def to_checkpoint_dict(state):
```
Comment (xibinliu) [inline src/maxtext/common/train_state_nnx.py:204]: state: nnx.Module

File: src/maxtext/common/checkpointing.py
Comment (xibinliu) [inline src/maxtext/common/checkpointing.py:336]: get param only model from nnx_params_abstract

Comment (xibinliu) [inline src/maxtext/common/checkpointing.py:350]: No need to use call _rebuild_nnx_with_values

Comment (xibinliu) [inline src/maxtext/common/checkpointing.py:1006]: No need to call this func. Merge NNX with linen logic
